File: 1054.distant-barcodes/solution.cpp
class Solution {
public:
    vector<int> rearrangeBarcodes(vector<int>& barcodes) {
        sort(barcodes.begin(), barcodes.end());
        vector<pair<int, int>> vec;
        int last = 0;
        int count = 0;
        for (int i : barcodes) {
            if (last != 0 && last != i)
            {
                vec.push_back(make_pair(count, last));
                count = 0;
            }
            last = i;
            count++;
        }
        vec.push_back(make_pair(count, last));
        sort(vec.begin(), vec.end());
        reverse(vec.begin(), vec.end());
        int index = 0;
        for (pair<int, int> &p : vec) {
            count = 0;
            while (count++ < p.first) {
                barcodes[index++] = p.second;
            }
        }
        vector<int> result(barcodes.size(), 0);
        index = -2;
        last = 0;
        for (int i : barcodes) {
            if (last != 0 && last != i) index++;
            else index += 2;
            if (index >= barcodes.size()) index = 0;
            while (result[index] != 0) index++;
            result[index] = i;
            last = i;
        }
        return result;
    }
};

File: 1054.distant-barcodes/solution.kt
import java.util.PriorityQueue

class Solution {
    fun rearrangeBarcodes(barcodes: IntArray): IntArray {
        val counts = HashMap<Int, Int>()
        for (i in barcodes)
            counts.compute(i) { _, v -> (v ?: 0) + 1 }
        val countsVal = PriorityQueue<Pair<Int, Int>> { o1, o2 -> o2.second.compareTo(o1.second) }
        for ((b, c) in counts)
            countsVal.add(b to c)
        var lastVal = 0
        return IntArray(barcodes.size) { _ ->
            val it = countsVal.iterator()
            var pair = it.next()
            if (pair.first == lastVal)
                pair = it.next()
            it.remove()
            if (pair.second > 1)
                countsVal.add(pair.first to pair.second - 1)
            lastVal = pair.first
            lastVal
        }
    }
}
